cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll

```diff
@@ -90,6 +79,52 @@ class PureStrFunction extends AliasFunction, ArrayFunction, TaintFunction, SideE
   }
 }
 
+class StrLenFunction extends AliasFunction, ArrayFunction, SideEffectFunction {
+  StrLenFunction() {
+    exists(string name |
+      hasGlobalOrStdName(name) and
+      (
+        name = "strlen" or
+        name = "strnlen" or
+        name = "wcslen"
+      )
+      or
+      hasGlobalName(name) and
+      (
+        name = "_mbslen" or
+        name = "_mbslen_l" or
+        name = "_mbstrlen" or
+        name = "_mbstrlen_l"
+      )
+    )
+  }
+
+  override predicate hasArrayInput(int bufParam) {
+    getParameter(bufParam).getUnspecifiedType() instanceof PointerType
+  }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) {
+    getParameter(bufParam).getUnspecifiedType() instanceof PointerType
+  }
+
+  override predicate parameterNeverEscapes(int i) {
+    getParameter(i).getUnspecifiedType() instanceof PointerType
+  }
+
+  override predicate parameterEscapesOnlyViaReturn(int i) { none() }
+
+  override predicate parameterIsAlwaysReturned(int i) { none() }
+
+  override predicate hasOnlySpecificReadSideEffects() { any() }
+
+  override predicate hasOnlySpecificWriteSideEffects() { any() }
+
+  override predicate hasSpecificReadSideEffect(ParameterIndex i, boolean buffer) {
+    getParameter(i).getUnspecifiedType() instanceof PointerType and
+    buffer = true
+  }
```

Comment:
Do you see a good alternative to repeating all these predicates? As I understand it, all we want is to exclude `strlen` and friends from the taint rule. Could that be done with a subclass that just overrides the taint predicate to `none()`? It might require repeating part of the charpred, but maybe that can be pulled out to a private predicate or class.

Reply:
We could certainly avoid the repetition with this strategy, but I think there will be a cost in comprehensibility.  I consider the repeated predicates here to have a very low mental cost, because this kind of repetition is expected across the models library and the repeated predicate bodies themselves are either trivial or very simple.  Like rows in a database.

If we _do_ make this change, I'd definitely want to add the helper predicate for the `StrLenFunction` charpred, because otherwise we'd be solving repetition in one place by adding repetition in another.  That means we'd be adding a helper predicate and complicating the class hierarchy.  Probably slightly fewer lines of code, but arguably harder to read overall.

Another approach for improving things here might be to add helpful tools / defaults to the abstract classes in the model interfaces, so that this stuff can be expressed more concisely in the most common cases (e.g. having the three `AliasFunction` predicates default to `none()` so that you only need to override the one(s) you want).  That will make the models more concise, though potentially mistakes might be harder to spot?

Reply:
I'm convinced by your argument that the intended way to use `models` is to inherit from all relevant interface classes every time. If that's too much boiler-plate, we should solve it everywhere and not just for `strlen`.